kiwicopple/serverless-postgres
# Add Fly Volume to persist data between Machine restarts

OUTDATED:  @pragmaticanon  beat me to it with this PR: https://github.com/kiwicopple/serverless-postgres/pull/3.  We can close this one and I can open a smaller one for the minor changes I included.

### Original PR notes:

This PR addresses the data loss that occurs in Fly when a Machine stops and is restarted.  This is accomplished by mounting a Fly Volume for the Postgres data we want to persist.  This addresses item 1 in this Issue: https://github.com/kiwicopple/serverless-postgres/issues/2

Even though we have the S3 storage with OrioleDB, we still want to preserve the local filesystem data between restarts.  Otherwise, we'd probably need to reinitialize the database from scratch from S3 on each start (likely using the python script discussed in the linked issue).

This line in our config sets the location of our data directory to `/var/lib/postgresql/data`:
https://github.com/kiwicopple/serverless-postgres/blob/418dbd348a9ffaf3ba515a2110f551f0d399cdcd/oriole/postgresql/postgresql.conf#L42

These changes mount the Fly Volume to the parent directory (`/var/lib/postgresql/`) since we can't mount the `data` directory directly (fails on initialization since it isn't empty - error due to 'lost and found' directory created at mount point).

Testing:
- Start a new PG instance in Fly
- Set up OrioleDB and add a table and some data
- Stop the Fly Machine
- Start the Fly Machine
- Reconnect to the database
- Confirm table/rows are still present

Considerations:
Initial testing indicates that this mount point will work, but we may later find we want to mount something higher up the tree.

Minor changes included:
- Added `.idea` to gitignore.  I'm using the database functionality in an IntelliJ IDE.
- Clarified what the expected data is in one environment variable (the bucket name).
